DocumentsFromSnapshotMigration/src/test/java/org/opensearch/migrations/bulkload/LeaseExpirationTest.java

```diff
@@ -0,0 +1,273 @@
+package org.opensearch.migrations.bulkload;
+
+import java.io.File;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+
+import org.opensearch.migrations.CreateSnapshot;
+import org.opensearch.migrations.bulkload.common.OpenSearchClient;
+import org.opensearch.migrations.bulkload.common.http.ConnectionContextTestParams;
+import org.opensearch.migrations.bulkload.framework.SearchClusterContainer;
+import org.opensearch.migrations.bulkload.http.ClusterOperations;
+import org.opensearch.migrations.data.IndexOptions;
+import org.opensearch.migrations.data.WorkloadGenerator;
+import org.opensearch.migrations.data.WorkloadOptions;
+import org.opensearch.migrations.data.workloads.Workloads;
+import org.opensearch.migrations.reindexer.tracing.DocumentMigrationTestContext;
+import org.opensearch.migrations.snapshot.creation.tracing.SnapshotTestContext;
+import org.opensearch.migrations.testutils.ToxiProxyWrapper;
+
+import eu.rekawek.toxiproxy.model.ToxicDirection;
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+import org.jetbrains.annotations.NotNull;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Tag;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+import org.testcontainers.containers.Network;
+
+/**
+ * TODO - the code in this test was lifted from ProcessLifecycleTest.java
+ * Some of the functionality and code are shared between the two and should be refactored.
+ */
+@Slf4j
+public class LeaseExpirationTest extends SourceTestBase {
+
+    public static final String TARGET_DOCKER_HOSTNAME = "target";
+    public static final String SNAPSHOT_NAME = "test_snapshot";
+
+    @AllArgsConstructor
+    @Getter
+    private static class RunData {
+        Path tempDirSnapshot;
+        Path tempDirLucene;
+        ToxiProxyWrapper proxyContainer;
+    }
+
+    @Tag("isolatedTest")
+    @ParameterizedTest
+    @ValueSource(booleans = {true, false})
+    public void testProcessExitsAsExpected(boolean forceMoreSegments) {
+        // Sending 5 docs per request with 4 requests concurrently with each taking 0.125 second is 160 docs/sec
+        // will process 9760 docs in 61 seconds. With 20s lease duration, expect to be finished in 4 leases.
+        // This is ensured with the toxiproxy settings, the migration should not be able to be completed
+        // faster, but with a heavily loaded test environment, may be slower which is why this is marked as
+        // isolated.
+        // 2 Shards, for each shard, expect three status code 2 and one status code 0 (4 leases)
```

Comment:
I suspect this tuning is going to break down between different hardware on dev machines and build labs, we might need to find another way to manage this work.

Reply:
Not exactly, this is because of toxiproxy and running as isolated.
By increasing the lease and the artificial delay, we can make it more consistent under system stress. What I added here was a buffer from the minimum that worked on my laptop and it worked first attempt on GHA.

Reply:
e.g. Say this entire workload can be done in 10s on jenkins and 15s on my laptop and 20 s on a slower laptop, by adding 3 minutes of artificial delay, it will take 3:10, 3:15, 3:20, which makes it possible to consistently assert on an expected time range (e.g. between 3:00 - 3:30)
(asserting on number of leases is roughly equivalent to the time range)

Reply:
Thanks for including the thought process you are using here - my spider-sense is tingling but lets not take any additional action without a specific issue.